### PR TITLE
Implements Shibboleth into the application (#10).

### PR DESCRIPTION
**Pull request by bwatson78**
- .rubocop.yml: excludes already existing controller and spec files from length requirements.
- Gemfile*: adds the Shibboleth Omniauth Gem.
- app/controllers/catalog_controller.rb: pulls authentication key from the `uid`.
- app/controllers/omniauth_controller.rb and omniauth_callbacks_controller.rb: wires omniauth controllers appropriately.
- app/models/auth_config.rb: allows database-driven login in the development/text environments.
- app/models/user.rb: configures the omniauth/devise gems for the new provider, as well as providing a lookup tool.
- app/views/devise/sessions/new.html.erb: stubs our own login page.
- config/initializers/devise.rb: sets the details for operating shibboleth.
- config/routes.rb: reroutes login through Shibboleth.
- db/*: migrates the new, needed columns for users.
- spec/*: tests the functionality of shibboleth.

### CODE REVIEW

File: app/controllers/omniauth_callbacks_controller.rb [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# frozen_string_literal: true
+class OmniauthCallbacksController < Devise::OmniauthCallbacksController
+  def shibboleth
+    Rails.logger.debug "OmniauthCallbacksController#shibboleth: request.env['omniauth.auth']: #{request.env['omniauth.auth']}"
+    # had to create the `from_omniauth(auth_hash)` class method on our User model
+    @user = User.from_omniauth(request.env["omniauth.auth"])
+    set_flash_message :notice, :success, kind: "Shibboleth"
+    sign_in_and_redirect @user
```
Comment (devanshu-m) [inline app/controllers/omniauth_callbacks_controller.rb:8]: Can you add a condition here to check if the user is persisted, and only then sign them in? When logging into shib, we expect a user object to already exist in the database, please correct me if this is not a requirement for this app. Refer: https://github.com/emory-libraries/dlp-curate/blob/main/app/controllers/omniauth_callbacks_controller.rb#L7

File: app/models/user.rb [hunk 2]
```diff
@@ -2,15 +2,32 @@
...
+    uid
+  end
+
+  # When a user authenticates via shibboleth, find their User object or make
+  # a new one. Populate it with data we get from shibboleth.
+  # @param [OmniAuth::AuthHash] auth
+  def self.from_omniauth(auth)
+    Rails.logger.debug "auth = #{auth.inspect}"
+    # Uncomment the debugger above to capture what a shib auth object looks like for testing
+    user = where(provider: auth.provider, uid: auth.info.uid).first_or_create
```
Comment (devanshu-m) [inline app/models/user.rb:26]: Are we creating users when logging in via shibboleth? If not, we only do a hard `find_by` else error out to say `user doesn't exist` Please refer: https://github.com/emory-libraries/dlp-lux/blob/main/app/models/user.rb#L35

**Review by devanshu-m — COMMENTED** [review]
A couple of comments to make sure I understand requirements correctly.

Last question: Are we not saving ppid for a user in this app? I don't see the need to, only confirming.

**Comment by devanshu-m** [comment]
Great work, @bwatson78

File: app/controllers/omniauth_callbacks_controller.rb (hunk 1 above)
Comment (bwatson78) [reply app/controllers/omniauth_callbacks_controller.rb:8]: This is a direct copy of the Lux method, which is what I've been told we're mirroring for Blacklight.

File: app/models/user.rb (hunk 2 above)
Comment (bwatson78) [reply app/models/user.rb:26]: This applies in the same way to the above comment.

Comment (devanshu-m) [reply app/models/user.rb:26]: I was referring to lux in my comment. Please take a look at the line I have referred (which is from lux).

File: app/controllers/omniauth_callbacks_controller.rb (hunk 1 above)
Comment (devanshu-m) [reply app/controllers/omniauth_callbacks_controller.rb:8]: Correct, but checking if a user exists before signing them in is an improvement which would help us with better error handling. We can skip making the change for now and come back to it if the need arises.

Comment (bwatson78) [reply app/controllers/omniauth_callbacks_controller.rb:8]: No, I'm happy making the changes, it's just that I saw you referring to Curate, and the request was to emulate Lux. Will jump on the changes.

Comment (bwatson78) [reply app/controllers/omniauth_callbacks_controller.rb]: Changes have been pushed into the branch.

File: app/models/user.rb (hunk 2 above)
Comment (bwatson78) [reply app/models/user.rb:26]: I spoke earlier to @rotated8 about going further than DCE's initial Shib integration that they did in Lux, and he said that the main intent of the ticket was just to get the Shibboleth login working, and to ignore any refinement or guessing of PO requirements.

Comment (devanshu-m) [reply app/models/user.rb:26]: I wasn't aware of this, thanks for letting me know. I was looking at what lux had as a final implementation for shib. My earlier comment doesn't make sense in this situation then, I apologize.

**Review by devanshu-m — APPROVED** [review]

File: app/models/user.rb (hunk 2 above)
Comment (bwatson78) [reply app/models/user.rb:26]: No, no apologies needed. I should have documented the details of my conversation with Collin here before pushing work at all, so I'm the one to apologize--and I do.

**Pull request merged**